commit 7e8e21df0c9aa39278e994b05540b69920201b32
Author: Nikita Popov <nikic@php.net>
Date:   Sat Jan 4 01:22:14 2014 +0100

    Fix bug #66286: Incorrect object comparison with inheritance
    
    std_compare_objects immidiately returned 0 if the property tables
    of both objects contain NULL at some index. Thus it would report
    objects as equal even though properties following after that
    differ.

--- /dev/null
+++ php5/Zend/tests/bug66286.phpt
@@ -0,0 +1,26 @@
+--TEST--
+Bug #66286: Incorrect object comparison with inheritance
+--FILE--
+<?php
+
+abstract class first {
+    protected $someArray = array();
+}     
+
+class second extends first {    
+    protected $someArray = array();        
+    protected $someValue = null;
+
+    public function __construct($someValue) {
+        $this->someValue = $someValue;
+    }
+}
+
+$objFirst = new second('123');       
+$objSecond = new second('321');       
+
+var_dump ($objFirst == $objSecond);    
+
+?>
+--EXPECT--
+bool(false)
--- php5.orig/Zend/zend_object_handlers.c
+++ php5/Zend/zend_object_handlers.c
@@ -1376,10 +1376,6 @@ static int zend_std_compare_objects(zval
 					Z_OBJ_UNPROTECT_RECURSION(o1);
 					Z_OBJ_UNPROTECT_RECURSION(o2);
 					return 1;
-				} else {
-					Z_OBJ_UNPROTECT_RECURSION(o1);
-					Z_OBJ_UNPROTECT_RECURSION(o2);
-					return 0;
 				}
 			}
 		}
